### PR TITLE
Improve lint import warnings

### PR DESCRIPTION
**Pull request by matthew-hilty**
Two main changes:
1. Add an import declaration's optional qualifier as a parameter to `UnusedImport`.
    - Previously, warnings didn't distinguish between import declarations from the same module.
      Code like the following
      ```purescript
      import A.B (x) -- `x` is used.
      import A.B (y) as C -- `y` is not used.
      ```
      would induce a warning like `The import of module A.B is redundant` even though only the qualified import declaration `C` is actually redundant.
    - The warning now would be `The import of module A.B (qualified as C) is redundant`.
2. Include kind imports among all imports considered by the linter for determining `UnusedImport` and `UnusedExplicitImport` warnings.
    - Previously, kind imports were ignored.
      The linter wouldn't emit any warnings for code like the following.
      ```purescript
      import A.B (kind K) -- `kind K` is not used.
      ```
    - And the linter, disregarding `kind K`, would emit an `UnusedImport` instead of an `UnusedExplicitImport` for code like the following.
      ```purescript
      import A.B (x, kind K) -- `x` is not used, but `kind K` is.
      ```

Issues #1647, #2339, #1897, #1647, #1842, #1670 are similar to this issue, but they relate to warnings/errors about types or exports rather than about imports.

### CODE REVIEW

File: src/Language/PureScript/Sugar/Names/Imports.hs [hunk 1]
```diff
@@ -32,10 +32,9 @@ findImports
   -> M.Map ModuleName [ImportDef]
 findImports = foldr go M.empty
   where
-  go (ImportDeclaration (pos, _) mn typ qual) result =
-    let imp = (pos, typ, qual)
-    in M.insert mn (maybe [imp] (imp :) (mn `M.lookup` result)) result
-  go _ result = result
+  go (ImportDeclaration (pos, _) mn typ qual) =
+    M.alter (return . ((pos, typ, qual) :) . fromMaybe []) mn
+  go _ = id
```
Comment (hdgarrood) [inline src/Language/PureScript/Sugar/Names/Imports.hs]: Am I correct in thinking the change in this file is just a refactor which shouldn't change behaviour?

File: src/Language/PureScript/Errors.hs [hunk 2]
```diff
@@ -911,8 +911,13 @@ prettyPrintSingleError (PPEOptions codeColor full level showDocs relPath) e = fl
...
             ]
-    renderSimpleErrorMessage (UnusedImport name) =
-      line $ "The import of module " <> markCode (runModuleName name) <> " is redundant"
+
+    renderSimpleErrorMessage (UnusedImport mn qualifier) =
+      let
+        mark = markCode . runModuleName
+        qual' q = " (qualified as " <> mark q <> ")"
+        qual = maybe "" qual'
+      in line $ "The import of module " <> mark mn <> qual qualifier <> " is redundant"
```
Comment (hdgarrood) [inline src/Language/PureScript/Errors.hs]: Bit of a nit but may I suggest copying GHC and saying "The qualified import of [module] as [qualified-name] is redundant"?

File: src/Language/PureScript/Sugar/Names/Imports.hs (hunk 1 above)
Comment (matthew-hilty) [reply src/Language/PureScript/Sugar/Names/Imports.hs]: Yeah. I think Gary's advice from a different pull request was still in mind when I made this change. It's meant to be just a simple refactor like you say, but maybe I should undo this patch so everything here is related directly to linting. I don't mind either way.

File: src/Language/PureScript/Errors.hs (hunk 2 above)
Comment (matthew-hilty) [reply src/Language/PureScript/Errors.hs]: Sure, that seem likes a good idea. I'll change the message, and I'll probably be able to push up the changes by tonight.

(By the way, please, pick as many nits as you want. I appreciate any tips/advice you guys have for me.)

File: src/Language/PureScript/Sugar/Names/Imports.hs (hunk 1 above)
Comment (hdgarrood) [reply src/Language/PureScript/Sugar/Names/Imports.hs]: Happy to keep this in here - I was just wondering.

**Comment by hdgarrood** [comment]
@matthew-hilty looks like your rebase has gone a bit wrong. Would you mind trying again?

**Comment by thomashoneyman** [comment]
Anything stopping this one from going in? Looks like it's up to date and passing tests, and I encountered a situation today at work where this would have been helpful.

**Comment by hdgarrood** [comment]
Nope! Thanks again @matthew-hilty

**Pull request merged**